mdbenchmark/utils.py

```diff
@@ -125,3 +130,85 @@ def generate_output_name(extension):
     date_time = dt.datetime.now().strftime("%Y-%m-%d_%H%M%S")
     out = "{}.{}".format(date_time, extension)
     return out
+
+def DataFrameFromBundle(bundle):
+    """Generates a data frame from a datreant bundle."""
+    df = pd.DataFrame(
+        columns=["module", "nodes", "ns/day", "run time [min]", "gpu", "host", "ncores"]
+    )
+
+    for i, sim in enumerate(bundle):
+        # older versions wrote a version category. This ensures backwards compatibility
+        if "module" in sim.categories:
+            module = sim.categories["module"]
+        else:
+            module = sim.categories["version"]
+        # call the engine specific analysis functions
+        engine = detect_md_engine(module)
+        df.loc[i] = utils.analyze_run(engine=engine, sim=sim)
+
+    if df.empty:
+        console.error("There is no data for the given path.")
+
+    # Sort values by `nodes`
+    df = df.sort_values(
+        ["host", "module", "run time [min]", "gpu", "nodes"]
+    ).reset_index(drop=True)
+
+    return df
+
+
+def ConsolidateDataFrame(df):
+    """Edits a dataframe and consolidates the output.
+       Requires the previously generated dataframe.
+       Returns a newly formatted dataframe.
+    """
+    df_short = pd.DataFrame(
+        columns=["module", "nodes", "run time [min]", "host", "gpu"]
+    )
+
+    groupby = ["module", "host", "gpu"]
+    gb = df.groupby(groupby)
+
+    i = 0
+    for key, df in gb:
+
+        node_print_output = []
+        node_groups = group_consecutives(df['nodes'].tolist())
+
+        for node_g in node_groups:
+            if len(node_g) is 1:
+                 node_print_output.append(node_g[0])
+            elif len(node_g) is not 1:
+                 node_print_output.append(str(node_g[0]) + "-" + str(node_g[-1]))
+
+        values = ','.join(str(v) for v in node_print_output)
+        df_short.loc[i] = (key[0], values, df["run time [min]"].iloc[0], key[1], key[2])
+        i += 1
+
+    return df_short
+
+
+def PrintDataFrame(df):
+    """Print a nicely formatted shortened dateframe."""
+
+    return
+
+
+def group_consecutives(vals, step=1):
+    """Return list of consecutive lists of numbers from vals (number list).
+       Adapted from code found on stack overflow.
```

Comment:
Please provide a source to the StackOverflow answer (not the question). The SO license asks for this.

https://stackoverflow.blog/2009/06/25/attribution-required/